### PR TITLE
feat(hacbs-994): Change CI to use catalog dir

### PR DESCRIPTION
**Pull request by parasense**
Signed-off-by: Jon Disnard <jdisnard@redhat.com>

### CODE REVIEW

**Comment by parasense** [comment]
@davidmogar @theflockers @johnbieren 
PTAL

File: .github/scripts/tkn_push_bundles.sh [hunk 1]
```diff
@@ -14,18 +14,24 @@ printf 'bundle_dirs: %s\n' "${!bundle_dirs[@]}"
...
-  printf '* Bundle_dir: %s\n' $BUNDLE_DIR
+  RESOURCE_DIR=${BUNDLE_DIR%/*}
+  RESOURCE_TYPE=${RESOURCE_DIR#*/}
+  printf '* Version dir: %s\n' $VERSION_DIR
+  printf '* Bundle dir: %s\n' $BUNDLE_DIR
   printf '* Bundle name: %s\n' $BUNDLE_NAME
-  for file in $BUNDLE_DIR/*.yaml
+  printf '* Resource dir: %s\n' $RESOURCE_DIR
+  printf '* Resource type: %s\n' $RESOURCE_TYPE
+  for file in $VERSION_DIR/*.yaml
```
Comment (parasense) [inline .github/scripts/tkn_push_bundles.sh:34]: This could probably change.
In Tekton Catalog the YAML file **MUST** be the exact same as the parent folder.
For example:
`catalog/task/foo  --> catalog/task/foo/0.1/foo.yaml`

So no need to glob anymore. The idea of multiple YAMLs would end here.

Thoughts?

File: .github/scripts/tkn_push_bundles.sh [hunk 2]
```diff
@@ -14,18 +14,24 @@ printf 'bundle_dirs: %s\n' "${!bundle_dirs[@]}"
 
 # Main loop
 printf 'Pushing any new/modified bundle definitions.\n'
-for BUNDLE_DIR in "${!bundle_dirs[@]}"
+for VERSION_DIR in "${!bundle_dirs[@]}"
```
Comment (johnbieren) [inline .github/scripts/tkn_push_bundles.sh:17]: this is confusing to me. If they are version directories, let's rename the array as such. If they array is still holding bundle directories, then let's keep the variable name BUNDLE_DIR as is

File: .github/scripts/tkn_push_bundles.sh [hunk 3]
```diff
@@ -14,18 +14,24 @@ printf 'bundle_dirs: %s\n' "${!bundle_dirs[@]}"
...
+    printf 'Skipping non-existing bundle-path: %s\n' "$VERSION_DIR"
     continue
   fi
   BUNDLE_FILES=()
+  BUNDLE_DIR=${VERSION_DIR%/*}
   BUNDLE_NAME=${BUNDLE_DIR##*/}
-  printf '* Bundle_dir: %s\n' $BUNDLE_DIR
+  RESOURCE_DIR=${BUNDLE_DIR%/*}
+  RESOURCE_TYPE=${RESOURCE_DIR#*/}
+  printf '* Version dir: %s\n' $VERSION_DIR
```
Comment (johnbieren) [inline .github/scripts/tkn_push_bundles.sh:29]: all these statements seem a little excessive to me. If we want to print so much data, can we just say like "processing yaml file a/b/c/d/X.yaml" or something and cut down on these variables? It is hard to follow imo

File: .github/workflows/tekton_bundle_push.yaml [hunk 4]
```diff
@@ -3,7 +3,10 @@ name: Tekton Bundle Push
 on:  # yamllint disable-line rule:truthy
   push:
     branches: ['main']
-    paths: ['definitions/*/*.yaml']
+    paths:
+      - 'catalog/*/*/*/*.yaml' # E.G. catalog/pipeline/release/0.2/release.yaml
+      - '!catalog/*/*/*/samples/**' # negated paths
```
[the remark this answers is not in the document]
Comment (parasense) [reply .github/workflows/tekton_bundle_push.yaml:8]: There could be a "samples" directory, and if there is... then we do not want to trigger a bundle push on any changes happening there.

File: .github/scripts/tkn_push_bundles.sh (hunk 3 above)
Comment (parasense) [reply .github/scripts/tkn_push_bundles.sh]: Agreed. I've refactored those shell variables into positional-parameters.
I'm not sure it's any easier to read... I had to add a comment explaining what the positional parameters represent.
But the code is certainly more simplified from a runtime perspective, avoiding repetitive string parsing on each successive part of the YAML file path.

File: .github/scripts/tkn_push_bundles.sh (hunk 2 above)
Comment (parasense) [reply .github/scripts/tkn_push_bundles.sh]: Good idea. The latest commit changes the variable name to be more clear.

File: .github/scripts/tkn_push_bundles.sh (hunk 1 above)
Comment (parasense) [reply .github/scripts/tkn_push_bundles.sh]: I went ahead and made this change. 
There cannot be multiple YAML files anymore, per Tekton Catalog spec.

File: .github/scripts/tkn_push_bundles.sh (hunk 3 above)
Comment (johnbieren) [reply .github/scripts/tkn_push_bundles.sh]: I like it a lot better now and think the comment you added is very useful

File: .github/workflows/tekton_bundle_push.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (johnbieren) [reply .github/workflows/tekton_bundle_push.yaml:8]: I suppose. I guess my thought is if someone added a samples directory, they should also change the CI to negate the path in the same PR. But it is fine to do it now if a directory called samples is some common thing in the catalog structure that we may add in the future

**Review by johnbieren — APPROVED** [review]
I am good with the code but I think it would be good to get at least one other from the team to give it a look

**Review by theflockers — APPROVED** [review]
The shell code looks good for me. I would suggest to the team to use the "ShellCheck" extension on VScode as it checks the code style with some standards. Some bash styling stuff is hard to see so using it makes it easier to standardize the way we write shell scripts.

**Comment by parasense** [comment]
> The shell code looks good for me. I would suggest to the team to use the "ShellCheck" extension on VScode as it checks the code style with some standards. Some bash styling stuff is hard to see so using it makes it easier to standardize the way we write shell scripts.

Thanks, I installed "ShellCheck" extension, and fixed one of the findings, but ignored another finding which was false-positive. Namely `ShellCheck` is really pedantic about quoting variables, presumably to avoid word-splitting mistakes... but this script uses naked word-splitting as a feature... and so it goes.

I've seen other team use ShellCheck as part of their CI pipeline/workflow. Maybe someday we can do that too.

Also, while reviewing the script with ShellCheck, I noticed a mistake I hadn't caught before. So the latest commit has some minimal extra changes. If you care to take a look, I'd appreciate the extra set of eyes. Thanks!
